### PR TITLE
Bug 170 fix the missing services resolved event sending

### LINKED ISSUE #170: Services that were removed are reported as resolved.
Sometimes a service that was removed gets reported as resolved at some later point. I think I have a fix for this, but I need you guys to take a look. Basically in the method "updateRecord" in the JmDNSImpl when a serviceInfo is about to be removed it is not removed from the listeners list which is basically a list of ServiceInfo instances. I believe that afterwards this ServiceInfo object is not removed at all and somtimes this results in sending Resolved events for this service info.

### PR DESCRIPTION
**Pull request by IvanBoychevMarinov**
As I described in the issue itself I'm getting resolved events for services already removed. I think this is because the listeners collection is not cleared but I need you guys to verify and confirm this solution is not breaking anything else.

Fixes #170

### CODE REVIEW

**Comment by IvanBoychevMarinov** [comment]
Sorry, about the formatting. It was done automatically and I thought that it is not bad thing, so I decided to leave it this way. I will revert these right away.

About the solution... It really seems like it is doing the right thing but I was afraid I might be missing some other aspect that might got broken by this fix. 

I will remove the formatting and try to explain the fix afterwards .

**Comment by IvanBoychevMarinov** [comment]
Ok, I think I was able to revert the formatting. Now there are few util classes used in the test, the test itslef and the fix. 

It seems that the javax.jmdns.impl.tasks.resolver.ServiceInfoResolver adds a "listener" to the JmDNSImpl. This listener is actually an instance of the "ServiceInfoImpl". This listener might get "removed" in the "cancel" method, but this won't happen if the ServceInfo is "persistent" which, according to the JavaDoc means that

>  _Returns <code>true</code> if ServiceListener.resolveService will be called whenever new new information is received._

 which seems ok until the ServiceInfo is there. I really think this "listener" aka the ServiceInfoImpl has to be removed when the ServiceInfo is removed. I have traced the chain of starting a service resolver and it seems a service resolver will be started if the service is added once again and it makes me feel that this is the right thing to do in this case.

Oh and the issue itself is caused by obsolete "listener" aka ServiceInfoImpl instance that invokes dns.handleServiceResolved in its updateRecord method.

**Comment by IvanBoychevMarinov** [comment]
@kaikreuzer, can you please review the changes. The biggest change is in the JmdnsImpl file. The rest are just two helpful test classes and a test method in the JmdnsImplTest.

**Comment by IvanBoychevMarinov** [comment]
@ViToni I saw you were the last one that modified the updateRecord method in the JmDNSImpl. Can you please take a look. 
To describe the issue in short, since the listeners were never stopped, they are not removed and thus they are constantly notified for service resolved even if those services are removed. You can see my previous comments for further info. The other changes are just unit tests demonstrating the behvaiour.

File: src/main/java/javax/jmdns/impl/JmDNSImpl.java [hunk 1]
```diff
@@ -1018,12 +1018,15 @@ public void registerService(ServiceInfo infoAbstract) throws IOException {
         info.addAddress(_localHost.getInet4Address());
         info.addAddress(_localHost.getInet6Address());
 
+        this.waitForAnnounced(DNSConstants.SERVICE_INFO_TIMEOUT);
```
Comment (kaikreuzer) [inline src/main/java/javax/jmdns/impl/JmDNSImpl.java:1021]: How is this related to the bug?
Isn't this dangerous here as it blocks the method from returning?

File: src/main/java/javax/jmdns/impl/JmDNSImpl.java [hunk 2]
```diff
@@ -1018,12 +1018,15 @@ public void registerService(ServiceInfo infoAbstract) throws IOException {
...
+
         this.makeServiceNameUnique(info);
         while (_services.putIfAbsent(info.getKey(), info) != null) {
             this.makeServiceNameUnique(info);
         }
 
         this.startProber();
+        info.waitForAnnounced(DNSConstants.SERVICE_INFO_TIMEOUT);
```
Comment (kaikreuzer) [inline src/main/java/javax/jmdns/impl/JmDNSImpl.java:1029]: Why a second time here?

File: src/test/java/javax/jmdns/impl/EventContainer.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import javax.jmdns.ServiceEvent;
+
+public class EventContainer {
+
+    private final Map<String, ServiceEvent> events = new LinkedHashMap<String, ServiceEvent>();
+
+    private boolean                         eventReceived;
```
Comment (kaikreuzer) [inline src/test/java/javax/jmdns/impl/EventContainer.java:12]:
```suggestion
    private boolean eventReceived;
```

File: src/test/java/javax/jmdns/impl/JmDNSTest.java [hunk 4]
```diff
@@ -557,4 +558,45 @@ public void testListMyServiceWithoutLowerCase() throws IOException, InterruptedE
             if (registry != null) registry.close();
         }
     }
+
+    @Test
+    public void shouldNotNotifyServiceListenersForServiceResolveAfterServiceRemoved() throws UnknownHostException, IOException, InterruptedException {
+        String firstType = "_test._type1.local.";
+        String secondType = "_test._type2.local.";
+
+        JmmDNS jmmdns = JmmDNS.Factory.getInstance();
```
Comment (kaikreuzer) [inline src/test/java/javax/jmdns/impl/JmDNSTest.java:567]: We are here in the JmDNSTest class, not the JmmDNSTest class - I'd hence assume you should also create JmDNS instances here, right?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Hey @IvanBoychevMarinov,

Very sorry for having neglected that project for such a long time and your PR with it.
I'm just going through the open PRs right now and I hope you are still around and interested in following up on that one! Let me know, what you think wrt my comments below. Thanks!

File: src/main/java/javax/jmdns/impl/JmDNSImpl.java (hunk 1 above)
Comment (IvanBoychevMarinov) [reply src/main/java/javax/jmdns/impl/JmDNSImpl.java:1021]: Sorry, I will have to take a look to verify why is this shown as it is part of my commit. It is not. I presume it was there before my changes and I just moved it.

File: src/main/java/javax/jmdns/impl/JmDNSImpl.java (hunk 2 above)
Comment (IvanBoychevMarinov) [reply src/main/java/javax/jmdns/impl/JmDNSImpl.java:1029]: Sorry, I will have to take a look to verify why is this shown as it is part of my commit. It is not. I presume it was there before my changes and I just moved it.

File: src/test/java/javax/jmdns/impl/JmDNSTest.java (hunk 4 above)
Comment (IvanBoychevMarinov) [reply src/test/java/javax/jmdns/impl/JmDNSTest.java:567]: Sorry, for two years, I have to recall a lot of things. There a lot of things I do not remember. I will try to provide answers soon.

File: src/main/java/javax/jmdns/impl/JmDNSImpl.java (hunk 1 above)
Comment (IvanBoychevMarinov) [reply src/main/java/javax/jmdns/impl/JmDNSImpl.java:1021]: I did a quick check and this strings appears in two commits
* 9689c79aa270572a1f48878a52a93ccbe4aa3d9a -  Added timeout on waitForAnnounced
* fee0d4c27cdc159a1b4a0fa5883ddf562a696d00 -  Remove waits. Fixes #144 (#174)

I'm not sure why those waits got back in my commit, but I will remove it. This is the only non intended difference between my commit and fee0d4c27cdc159a1b4a0fa5883ddf562a696d00 As far as I remember this happened in some intermediate release.

File: src/test/java/javax/jmdns/impl/JmDNSTest.java (hunk 4 above)
Comment (IvanBoychevMarinov) [reply src/test/java/javax/jmdns/impl/JmDNSTest.java:567]: I just took a look and instances of the JmDNS are actually created few lines below in a for loop for each InetAddress.

**Comment by mkorwel** [comment]
Hi 
@kaikreuzer. What next with this PR? Are you going to merge it? Are you planning next release with it?
@IvanBoychevMarinov thanks for you changes. I tested it in my case and your fix solves my problem.

**Review by kaikreuzer — APPROVED** [review]
Many thanks and sorry for the very late processing. Will do a new release before the end of the month!

**Pull request merged**